[cocoa] Add Continue button on warning view

sysrqb opened the pull request:
#### c82e784731d3ce16ffdb4d3e9f3158618184760f
<pre>
[cocoa] Add Continue button on warning view
<a href="https://bugs.webkit.org/show_bug.cgi?id=277529">https://bugs.webkit.org/show_bug.cgi?id=277529</a>
<a href="https://rdar.apple.com/133037364">rdar://133037364</a>

Reviewed by Alex Christensen.

Adds support for a &quot;Continue&quot; button in the same place as the current &quot;Show
Details&quot; button when the warning is for a HTTPSNavigationFailure.

As part of this change, I&apos;m storing the BrowsingWarning&apos;s Data parameter and
providing an accessor for it so we can identify the desired warning.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _showWarningViewWithURL:title:warning:detailsWithLinks:completionHandler:]):

This patch now requires explicitly declaring the warning type.

* Source/WebKit/UIProcess/BrowsingWarning.h:
(WebKit::BrowsingWarning::create):
(WebKit::BrowsingWarning::data const):
* Source/WebKit/UIProcess/Cocoa/BrowsingWarningCocoa.mm:
(WebKit::BrowsingWarning::BrowsingWarning):
* Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
(colorForItem):
(makeButton):
(-[_WKWarningView addContent]):
(-[_WKWarningView continueClicked]):

Canonical link: <a href="https://commits.webkit.org/281796@main">https://commits.webkit.org/281796@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7efa22bc1dbadcacf300c60c99b48c2320eb82ca

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40345 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13562 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64919 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11533 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48021 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11808 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49290 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7995 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/63020 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37550 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52843 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30119 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34233 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10446 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/56066 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10350 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66649 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4930 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10185 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56659 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4952 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52799 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56849 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4084 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9183 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37234 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38328 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3673,7 +3673,7 @@ - (void)_showSafeBrowsingWarningWithURL:(NSURL *)url title:(NSString *)title war
 - (void)_showWarningViewWithURL:(NSURL *)url title:(NSString *)title warning:(NSString *)warning detailsWithLinks:(NSAttributedString *)details completionHandler:(void(^)(BOOL, NSURL *))completionHandler
```
I should've left this as `_showSafeBrowsingWarningWithURL` because it's only called from other `_showSafeBrowsingWarning` SPI.

achristensen07 replied:
I actually think it's better as _showWarningViewWithURL because I hope it is used for more diverse warnings in the future.

sysrqb replied:
Yeah, we'll just need to add another parameter for the warning data in that case.

achristensen07 commented on Source/WebKit/UIProcess/BrowsingWarning.h:
```diff
@@ -68,6 +68,7 @@ class BrowsingWarning : public RefCounted<BrowsingWarning> {
 #if PLATFORM(COCOA)
     RetainPtr<NSAttributedString> details() const { return m_details; }
 #endif
+    Data data() const { return m_data; }
```
const Data&

achristensen07 commented on Source/WebKit/UIProcess/BrowsingWarning.h:
```diff
@@ -87,6 +88,7 @@ class BrowsingWarning : public RefCounted<BrowsingWarning> {
 #if PLATFORM(COCOA)
     RetainPtr<NSAttributedString> m_details;
 #endif
+    Data m_data;
```
const Data

Commit:
[cocoa] Add Continue button on warning view
https://bugs.webkit.org/show_bug.cgi?id=277529
rdar://133037364

Reviewed by Alex Christensen.

Adds support for a "Continue" button in the same place as the current "Show
Details" button when the warning is for a HTTPSNavigationFailure.

As part of this change, I'm storing the BrowsingWarning's Data parameter and
providing an accessor for it so we can identify the desired warning.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _showWarningViewWithURL:title:warning:detailsWithLinks:completionHandler:]):

This patch now requires explicitly declaring the warning type.

* Source/WebKit/UIProcess/BrowsingWarning.h:
(WebKit::BrowsingWarning::create):
(WebKit::BrowsingWarning::data const):
* Source/WebKit/UIProcess/Cocoa/BrowsingWarningCocoa.mm:
(WebKit::BrowsingWarning::BrowsingWarning):
* Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
(colorForItem):
(makeButton):
(-[_WKWarningView addContent]):
(-[_WKWarningView continueClicked]):

Canonical link: https://commits.webkit.org/281796@main
